Feature/double hbone

Stevenjin8 opened the pull request:
More or less works other than gracefully shutting down connections.

I just need to fill in where to get remote workload info (e/w gateway and remote ztunnel).

The changes tls/cert validation are temporary as well as `localhost.yaml` and `local-dev-setup.sh`.

Stevenjin8 committed:
double hbone (no inner encryption) working!!

Stevenjin8 committed:
hbone-ish

Stevenjin8 commented on src/tls/certificate.rs:
```diff
@@ -227,6 +228,10 @@ impl WorkloadCertificate {
 
         let mut roots = RootCertStore::empty();
         roots.add_parsable_certificates(chain.iter().last().map(|c| c.der.clone()));
+        roots.add_parsable_certificates(vec![CertificateDer::from_pem_file(
```
Stuff in this file is a temporary workaround for certificates b/c I don't really know how they work.

Stevenjin8 commented on src/proxy/outbound.rs:
```diff
@@ -202,10 +207,43 @@ impl OutboundConnection {
+        let upgraded = Box::pin(self.send_hbone_request(remote_addr, req)).await?;
+
+        let upgraded = TokioH2Stream::new(upgraded);
+        let inner_workload = pool::WorkloadKey {
```
Its unclear to me the exact mechanisms we are going to use to get info of e/w gateways and final destination.

keithmattix replied:
The only info you need from the final destination is the hostname and the expected SANs. TBD on how to get that, but the e/w gateway info will come directly from the workload

Stevenjin8 commented on src/proxy/pool.rs:
```diff
@@ -71,18 +71,41 @@ struct PoolState {
     // to ensure we get unique poolkeys-per-new-conn, it is not a limit
     pool_global_conn_count: AtomicI32,
-    spawner: ConnSpawner,
+    pub spawner: ConnSpawner,
```
Need to hoist this out to keep member private. Any ideas other than a spawner() method?

Stevenjin8 committed:
cleanup

Stevenjin8 committed:
graceful shutdowns

Stevenjin8 committed:
Some cleanup

Stevenjin8 committed:
Add some auth/tls logic back in

Stevenjin8 committed:
inline double hbone code

Stevenjin8 committed:
Use correct(?) identities

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1435)

howardjohn committed:
Resync proto with istio/istio (#1434)

This fell out of date

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1437)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1439)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1440)

howardjohn committed:
metrics: add src/dst locality information to labels (#1441)

Understanding cross-zone/cross-region traffic is one of the most
critical pieces of observing a cluster, for a cost, reliability, and
latency standpoint. Lots of work has been done in this space from
standalone tools (https://github.com/polarsignals/kubezonnet/) to
enterprise upsells (I *think* Buoyant enterprise has this but not in
OSS?); we can provide this exact data trivially.

louiscryan committed:
Add new ztunnel images (#1443)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1444)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1445)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1451)

howardjohn committed:
bump dependencies (#1449)

* bump dependencies

Manually apply https://github.com/istio/ztunnel/pull/1448 to both
modules

* Update cert format

* Fix formatting as well

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1455)

howardjohn committed:
Bump dependencies (#1458)

* Bump dependencies

Few libraries making breaking changes in prep for Rust 2024, bump those.
Notably, the older `ctor` we were on emits warnings on later rust versions

* gen

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1459)

Winbobob committed:
Add xDS and CA headers/metadata support (#1452)

* Add support for XDS headers and CA headers from environment variables

* Add unit tests

* Linting

* Make headers use metadata vector

* Linting

* Fix

* More fix

---------

Co-authored-by: Fuyuan Bie <fuyuanbie@microsoft.com>

Winbobob committed:
Simplify the xDS/CA headers parse logic (#1460)

bleggett committed:
Mention this explicitly here since it isn't clear (#1462)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1463)

howardjohn committed:
Update to Rust edition 2024 (#1465)

`cargo fix --edition` gave a few areas where semantics *could* change.
None were relevant to us really, so I reverted them. The rest if `cargo
format` changes

howardjohn committed:
format with rust 1.85 (#1464)

howardjohn committed:
Add support for AWS-LC-RS TLS provider (#1466)

* Add support for AWS-LC-RS TLS provider

Related PR: https://github.com/istio/ztunnel/pull/1436/

Ref https://github.com/istio/ztunnel/issues/1323

* Revert default

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1468)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1469)

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1471)

howardjohn committed:
Add benchmarks for load balancing (#1470)

tedjpoole committed:
Add support for OpenSSL (#1436)

Signed-off-by: Ted Poole <tpoole@redhat.com>

howardjohn committed:
wds: implement capacity (#1472)

Ztunnel side impl of https://github.com/istio/istio/pull/55284

howardjohn committed:
tls: move to `tls-aws-lc` by default (#1467)

* tls: move to `tls-aws-lc` by default

* fuzz

bleggett committed:
Remove ConnClient, it's pointless now (#1453)

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

howardjohn committed:
Remove deadcode around BufReader (#1474)

This is never used for writing, so no need for a write impl

Stevenjin8 committed:
Merge branch 'master' into feature/double-hbone

howardjohn committed:
Drop service lookup that clones (#1475)

No need to clone, we can keep as an arc. We did this everywhere else but
missed this one

Stevenjin8 committed:
checkpoint

Stevenjin8 committed:
another checkpoint

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1477)

jaellio committed:
Add inbound support for svc hostname in authority (#1420)

* Support hostname in authority header

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* remove todos

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Use enum for host addr

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Add HboneAddress enum and fix tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update String usage to Strng and add comments

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update svc_hostname to return strng

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Fix build errors

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Replace assert with panic for test configuration failure

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Fix fmt errors

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Respond to PR feedback

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Respond to some PR feedback

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Move HboneAddress to proxy

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Intelligently select service from hostname and vip from service.
Implement TryFrom from uri to hboneAddress

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* - Fix hbone_addr borrow and only pass port
- Update hostname in proxy protocol to be hostname:port

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update build_inbound_request to accept Parts to improve
testability

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Add e2e for svc hostname

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* clean up e2es

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Remove unused import

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update service name to match hostname

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Rebase

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Add ipv4 check

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Fix fmt

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Rebase and remove proxy protocol svc hostname

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* remove failing test

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Translate service port to target port

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Fix unit tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* fix nit

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update error type

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Added support for named target ports

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Update namespaced tests to support multiple test cases

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Use arc service

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Reformat endpoint dereference

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

---------

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Stevenjin8 committed:
Use new xds

Stevenjin8 committed:
Merge branch 'master' into HEAD

Stevenjin8 committed:
Merge commit '59eedf0' into feature/double-hbone

dependabot[bot] committed:
Bump the cargo group across 2 directories with 2 updates (#1479)

Bumps the cargo group with 2 updates in the / directory: [ring](https://github.com/briansmith/ring) and [openssl](https://github.com/sfackler/rust-openssl).
Bumps the cargo group with 1 update in the /fuzz directory: [ring](https://github.com/briansmith/ring).


Updates `ring` from 0.17.11 to 0.17.12
- [Changelog](https://github.com/briansmith/ring/blob/main/RELEASES.md)
- [Commits](https://github.com/briansmith/ring/commits)

Updates `openssl` from 0.10.68 to 0.10.70
- [Release notes](https://github.com/sfackler/rust-openssl/releases)
- [Commits](https://github.com/sfackler/rust-openssl/compare/openssl-v0.10.68...openssl-v0.10.70)

Updates `ring` from 0.17.11 to 0.17.13
- [Changelog](https://github.com/briansmith/ring/blob/main/RELEASES.md)
- [Commits](https://github.com/briansmith/ring/commits)

---
updated-dependencies:
- dependency-name: ring
  dependency-type: direct:production
  dependency-group: cargo
- dependency-name: openssl
  dependency-type: direct:production
  dependency-group: cargo
- dependency-name: ring
  dependency-type: indirect
  dependency-group: cargo
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Stevenjin8 committed:
lint

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1480)

Stevenjin8 committed:
Fix type sizes

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1482)

Stevenjin8 committed:
Check workload potocol

Stevenjin8 committed:
wip

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1486)

Stevenjin8 committed:
Merge branch 'master' into feature/double-hbone

howardjohn committed:
misc refactors around inbound code (#1490)

* misc refactors around inbound code

Some of the initial choses we made have not really held up over the
tweaks we have made over the year. Clean things up a bit. This should
have no meaningful user impact, mostly a refactoring.

* check_from_network_gateway is not right; it is assuming we get the
  identity from the EW GW; in the modern multi-network scheme that is
never true as its forwarded e2e from the client (double hbone). Instead,
change the heuristic to be "if hostname, it is multi-network".
* Instead of trying to create the PROXY protocol handling in the main
  forwarding logic, refactor it into find_inbound_upstream.

* fix

* address comments

* fmt

istio-testing committed:
Automator: update common-files@master in istio/ztunnel@master (#1493)

Stevenjin8 committed:
initial impl of hostname

Stevenjin8 committed:
Merge branch 'master' into feature/double-hbone

Stevenjin8 committed:
tests passing

Stevenjin8 committed:
svc addressing

Stevenjin8 committed:
Check workload network when using network gateways
